### DISCUSSION TITLE
Using restrictions for elements (not species)

### DISCUSSION THREAD

**Discussion by Barbara-Esteves** [reaktoro/reaktoro#385] [Q&A]
### Description

Hello!
I am trying to construct a simple equilibrium calculation that considers 4 species in to liquid phases. I started the problem considering the tips from "Modeling of a VLL system. #377" discussion.
I believe that the assumptions for my problem are: 4 components - 2 phases + 2 = 4 degrees of freedom. In this case, I tried to define 4 state parameters (T, P, and 2 compositions) and **4 more restrictions for each of the components (total moles in both phases).**
The question: I was just able to set 2 restrictions (using EquilibriumSpecs) for 2 elements (for example, H2O that is the sum of H2O_l1 and H2O_l2). If I consider more elements, this is the message:

RuntimeError: ***ERROR***
The number of introduced p control variables (e.g., temperature, pressure, amounts of explicit titrants, custom variables) is 2. The number of introduced equation constraints is 3. These two numbers must be equal, otherwise the chemical equilibrium problem cannot be solved. Modify your chemical equilibrium specifications so that this requirement is satisfied.

If I consider only the 2 restrictions, the code runs, but it did not converge:

***WARNING***
The chemical equilibrium/kinetics calculation did not converge.
This can occur due to various factors, such as:
  1. Infeasible Problem Formulation: The problem you have defined lacks a feasible solution within the given constraints and assumptions.
  2. Out-of-Range Thermodynamic Conditions: The conditions you have specified for the system may fall outside the valid range supported by the thermodynamic models used.
  3. Numerical Instabilities: Convergence issues may arise from numerical problems during the execution of the chemical/kinetic equilibrium algorithm. Consider reporting the issue with a minimal reproducible example if you believe the algorithm is responsible for this issue.
Disable this warning message with Warnings.disable(906) in Python and Warnings::disable(906) in C++.

Is this a way that you can see to include the restriction of total moles of each of the 4 elements in my system?

Below, the code of my system (it does not include the part that creates this 4 species in the (l) and (liq) phases).

Thanks!!

### Example

liquid_1 = LiquidPhase("C4H9OH(l) C2H4O2(l) H2O(l) C6H12O2(l)")
liquid_1.set(ActivityModelPengRobinson())
liquid_2 = LiquidPhase("C4H9OH(liq) C2H4O2(liq) H2O(liq) C6H12O2(liq)")
liquid_2.set(ActivityModelPengRobinson())

phases = Phases(db)
phases.add(liquid_1)
phases.add(liquid_2)
system = ChemicalSystem(phases)

### Creating a chemical state object
state = ChemicalState(system)
# Degrees of freedom: 4 components - 2 phases + 2 = 4
state.setTemperature(80, "celsius")
state.pressure(70, "kPa")
state.set("C4H9OH(l)", 0.410, "mol")
state.set("C2H4O2(l)", 0.006, "mol")

restrictions = EquilibriumSpecs(system)
restrictions.elementAmount("C4H9OH")
restrictions.elementAmount("C2H4O2")

solver = EquilibriumSolver(restrictions)

cond = EquilibriumConditions(restrictions)
cond.elementAmount("C4H9OH", 0.4387, "mol")
cond.elementAmount("C2H4O2", 0.0232, "mol") 

res = solver.solve(state,cond)
assert res.succeeded()

### What is your operating system?

Linux

**Comment by allanleal** [comment]
Hi @Barbara-Esteves - thanks for initiating this discussion. Could you please attach the whole Python script so that I can quickly reproduce the issue here? 

From your code above, it seems the `EquilibriumSpecs` object is not been set adequately. For example, temperature and pressure are not set as constraints (in case these two variables are known). Something like below should be used:

```py
specs = EquilibriumSpecs(system)
specs.temperature()  
specs.pressure() 
```

Also, I think creating constraints for element/component amounts is not what you may actually want. Instead, you can create a `ChemicalState` object containing some initial species amounts that implicitly represents the total amounts of elements in the system. These will be conserved.

**Reply by Barbara-Esteves** [reply]
~~~py
from reaktoro import *

db = Database()
db.addElement(Element().withName("C4H9OH").withSymbol("C4H9OH").withMolarMass(0.0741216)) # kg/mol
# Here, we are creating only the liquid species          
db.addSpecies(Species().withName("C4H9OH(l)").withFormula("C4H9OH").withCharge(0.)
              .withElements(ElementalComposition([(db.elements().get("C4H9OH"), 1.0)]))
              .withAggregateState(AggregateState.Liquid)
              .withStandardGibbsEnergy(0.))
# Specifying critical properties
crprops = SubstanceCriticalProps("C4H9OH(l)")
crprops.setTemperature(563.10, "K")
crprops.setPressure(44.23e5, "Pa")
crprops.setAcentricFactor(0.5940)
CriticalProps.overwrite(crprops)

## Acetic acid, C2H4O2 (properties obtained in the Reaktoro database)
db.addElement(Element().withName("C2H4O2").withSymbol("C2H4O2").withMolarMass(0.06005196)) # kg/mol     
db.addSpecies(Species().withName("C2H4O2(l)").withFormula("C2H4O2").withCharge(0.)
              .withElements(ElementalComposition([(db.elements().get("C2H4O2"), 1.0)]))
              .withAggregateState(AggregateState.Liquid)
              .withStandardGibbsEnergy(0.))
crprops = SubstanceCriticalProps("C2H4O2(l)")
crprops.setTemperature(592.00, "K")
crprops.setPressure(57.86e5, "Pa")
crprops.setAcentricFactor(0.4670)
CriticalProps.overwrite(crprops)

## Water, H2O (properties obtained in the Reaktoro database)
db.addElement(Element().withName("H2O").withSymbol("H2O").withMolarMass(0.01801528)) # kg/mol     
db.addSpecies(Species().withName("H2O(l)").withFormula("H2O").withCharge(0.)
              .withElements(ElementalComposition([(db.elements().get("H2O"), 1.0)]))
              .withAggregateState(AggregateState.Liquid)
              .withStandardGibbsEnergy(0.))
crprops = SubstanceCriticalProps("H2O(l)")
crprops.setTemperature(647.10, "K")
crprops.setPressure(220.55e5, "Pa")
crprops.setAcentricFactor(0.3450)
CriticalProps.overwrite(crprops)

## 1-butyl acetate, C6H12O2
# https://webbook.nist.gov/cgi/cbook.cgi?ID=C123864&Units=SI&Mask=4#Thermo-Phase (Pc and Tc)
# https://www.chemeo.com/cid/48-904-5/Acetic-acid-butyl-ester (acentric factor)
db.addElement(Element().withName("C6H12O2").withSymbol("C6H12O2").withMolarMass(0.1161583)) # kg/mol     
db.addSpecies(Species().withName("C6H12O2(l)").withFormula("C6H12O2").withCharge(0.)
              .withElements(ElementalComposition([(db.elements().get("C6H12O2"), 1.0)]))
              .withAggregateState(AggregateState.Liquid)
              .withStandardGibbsEnergy(0.))
crprops = SubstanceCriticalProps("C6H12O2(l)")
crprops.setTemperature(575.4, "K")
crprops.setPressure(230.90, "bar")
crprops.setAcentricFactor(0.4170)
CriticalProps.overwrite(crprops)

db.addSpecies(db.species("C4H9OH(l)").clone().withName("C4H9OH(liq)").withAggregateState(AggregateState.Liquid))
db.addSpecies(db.species("C2H4O2(l)").clone().withName("C2H4O2(liq)").withAggregateState(AggregateState.Liquid))
db.addSpecies(db.species("H2O(l)").clone().withName("H2O(liq)").withAggregateState(AggregateState.Liquid))
db.addSpecies(db.species("C6H12O2(l)").clone().withName("C6H12O2(liq)").withAggregateState(AggregateState.Liquid))

### Third step: setting up phases in chemical system (BIPs = 0)
liquid_1 = LiquidPhase("C4H9OH(l) C2H4O2(l) H2O(l) C6H12O2(l)")
liquid_1.set(ActivityModelPengRobinson())
liquid_2 = LiquidPhase("C4H9OH(liq) C2H4O2(liq) H2O(liq) C6H12O2(liq)")
liquid_2.set(ActivityModelPengRobinson())

### Defining chemical system
phases = Phases(db)
phases.add(liquid_1)
phases.add(liquid_2)
system = ChemicalSystem(phases)

### Creating a chemical state object
state = ChemicalState(system)
# Degrees of freedom: 4 components - 2 phases + 2 = 4
state.setTemperature(80, "celsius")
state.pressure(70, "kPa")
state.set("C4H9OH", 0.410, "mol")
state.set("C2H4O2(l)", 0.006, "mol")
#state.set("C4H9OH", 0.4387, "mol")

# Restrictions: for the 4 components 
# EquilibriumRestrictions - The class used to define reactivity restrictions in a chemical equilibrium calculation
# EquilibriumConditions - The class used to define conditions to be satisfied at chemical equilibrium
restrictions = EquilibriumSpecs(system)
restrictions.elementAmount("C4H9OH")
restrictions.elementAmount("H2O")
#restrictions.elementAmount("C2H4O2")
#restrictions.elementAmount("C6H12O2")

solver = EquilibriumSolver(restrictions)

cond = EquilibriumConditions(restrictions)
cond.elementAmount("C4H9OH", 0.4387, "mol")
cond.elementAmount("H2O", 1.5317, "mol")
#cond.elementAmount("C2H4O2", 0.0232, "mol") 
#cond.elementAmount("C6H12O2", 0.0064, "mol")

res = solver.solve(state,cond)
assert res.succeeded()

print(state)  # State after equilibration
~~~

**Reply by Barbara-Esteves** [reply]
Hi, Allan! Thanks for helping!
Probably I got confused trying to set the total amount of elements in my system. 
But the suggestion of using ChemicalState to set the initial amount of elements (moles in the liquid phases 1 and 2) does not work. Just for the species! 
For example, "Could not find a species with name or index `C4H9OH` in the chemical system.".
Should I specify any other thing to use ChemicalState and "set" for elements (not species)?

And just to understand, I intend to calculate the amounts in each phase based in total amount of each element in the system, so I can compare my results with some equilibrium data from the literature.

Thanks!
Barbara

**Comment by allanleal** [comment]
Hi Barbara. It is not possible indeed to set the amount of elements in a `ChemicalState` object, as you rightly pointed out above. What I meant is that you can achieve this in an implicit manner. For example, consider the initial state below in disequilibrium:

~~~py
state.set("C4H9OH(l)" , 0.004, "mol")
state.set("C2H4O2(l)" , 0.005, "mol")
state.set("H2O(l)"    , 0.990, "mol")
state.set("C6H12O2(l)", 0.001, "mol")

state.set("C4H9OH(liq)" , 0.150, "mol")
state.set("C2H4O2(liq)" , 0.150, "mol")
state.set("H2O(liq)"    , 0.001, "mol")
state.set("C6H12O2(liq)", 0.700, "mol")
~~~

Note that this sets an state with both phases containing certain amounts of each species. The "element" H2O has a total amount of 0.990 + 0.001 = 1.000 mol, while element C6H12O2 has 0.701 mol. 

I'll try to run your code later. But I do think the above way would be more adequate than using `EquilibriumSpecs.elementAmount`. 

Please note also that Reaktoro has yet to be better adapted for such LLE problems with hydrocarbons. Proper initial guesses are needed because you are using the same EOS for the same phases. Some discussions about this can be found in #377 .
